### ISSUE TITLE
[css-typed-om] Clarify list-valued properties

### ISSUE THREAD

**Issue by andruud** [w3c/css-houdini-drafts#823]
If a property optionally takes a non-list as a value, is it a [list-valued property](https://drafts.css-houdini.org/css-typed-om-1/#list-valued-properties)?

Examples:

1. `counter-reset`
2. `transform`
3. `content`
4. Custom property with syntax `none | <length>+`

If such properties are list-valued properties, what should happen when you do e.g. `e.attributeStyleMap.append('counter-reset', 'none')`?

**Comment by tabatkins** [comment]
We haven't figured out the answer to this. :( I suspect this needs to be resolved by separating out the top-level *productions* of a property grammar into "list-valued" and "single-valued" categories, and throwing an error if you ever try to put a single-valued production into a list-valued mode. (Such as starting with "transform:none" and then appending a value, or vice versa.)

Maybe we special-case list-based interaction with the "empty" values that some properties, like `transform`, have, so you can *start* from that value and append list-valued values and have it work "as expected"?

**Comment by andruud** [comment]
Describing list-valuedness of productions rather than properties makes much more sense!

Re. special-casery of "empty" values, something more general might be needed. Custom properties may have custom "empty" values, e.g. `omitted | <number>+`, and it's possible to register a property with several different incompatible lists: `<length>+ | <url>+`. Perhaps `append` should behave as `set` whenever you are trying to append something that's incompatible with the current value.

Sidenote: the `<length>+ | <url>+`-case means that we have to reify the current value to check whether the `append` is allowed. (We may be attempting to append a list of urls to an existing list of lengths, or vice-versa).

**Comment by tabatkins** [comment]
> Perhaps append should behave as set whenever you are trying to append something that's incompatible with the current value.

This is a good point. I'll think on this, but it sounds promising.

> Sidenote: the `<length>+ | <url>+`-case means that we have to reify the current value to check whether the append is allowed. (We may be attempting to append a list of urls to an existing list of lengths, or vice-versa).

Well, not *reify*, right? That's the process of turning the value into a JS object. This can be checked purely with the internal values, right?

**Comment by tabatkins** [comment]
Actually, I think what I want is:

* grammars have their top-level productions (the things separated by `|`) classified as list-valued or single-valued. (A single TypedOM object always represents either an *entire* single-valued term, or one iteration of a list-valued term.)
* you can't ever `append()` a value that matches a single-valued production. This throws an error.
* If you `append()` a value that's a list-valued iteration:
	* if the existing values are of the same production, great, it appends as normal.
	* if the existing value is of a single-valued production, it *replaces* the existing value silently. (This magically handles cases like appending to `none`, but also applies to any other single keywords/etc.)
	* if the existing values are of a *different* list-valued production, maybe that's an error? Or maybe it should replace, I'm unsure.

**Comment by andruud** [comment]
> Well, not reify, right? That's the process of turning the value into a JS object. This can be checked purely with the internal values, right?

Right. (To figure out compatibility between existing and incoming values, we _do_ have to create JS objects internally in Blink at the moment, but you're right: it's an implementation detail).

> you can't ever append() a value that matches a single-valued production. This throws an error.

Does this mean the following is an error?

```
// Given <length> | <length-percentage>+
e.attributeStyleMap.set('--x', '1%');
e.attributeStyleMap.append('--x', '1px');
```

**Comment by tabatkins** [comment]
Ah, bad phrasing on my part. To handle ambiguous things, I guess need to check it against the existing list-valued things. That is, verify that what's already in the property matches one of the list-valued iterations, and that the thing you're appending matches one of the *same* list-valued iterations.
